internal/zinc-classpath/src/main/scala/sbt/internal/inc/ScalaInstance.scala

```diff
@@ -123,30 +132,45 @@ object ScalaInstance {
       }
     }
     val jars = provider.jars
-    val loader = provider.loader
     val libraryJar = findOrCrash(jars, "scala-library.jar")
     val compilerJar = findOrCrash(jars, "scala-compiler.jar")
-    new ScalaInstance(version, loader, libraryJar, compilerJar, jars, None)
+    // sbt launcher 1.0.3 will construct layered classloader. Use them if we find them.
+    // otherwise, construct layered loaders manually.
+    val (loader, loaderLibraryOnly) =
+      (try {
+        provider match {
+          case p: ScalaProvider2 @unchecked => Option((provider.loader, p.loaderLibraryOnly))
+        }
+      } catch {
+        case _: NoSuchMethodError => None
+      }) getOrElse {
+        val l = ClasspathUtilities.toLoader(Vector(libraryJar))
+        val c = scalaLoader(l)(jars.toVector filterNot { _ == libraryJar })
+        (c, l)
+      }
+    new ScalaInstance(version, loader, loaderLibraryOnly, libraryJar, compilerJar, jars, None)
   }
 
   def apply(scalaHome: File, launcher: xsbti.Launcher): ScalaInstance =
-    apply(scalaHome)(scalaLoader(launcher))
+    apply(scalaHome)(scalaLibraryLoader(launcher))
 
   def apply(scalaHome: File)(classLoader: List[File] => ClassLoader): ScalaInstance = {
     val all = allJars(scalaHome)
-    val loader = classLoader(all.toList)
     val library = libraryJar(scalaHome)
+    val loaderLibraryOnly = classLoader(List(library))
+    val loader = scalaLoader(loaderLibraryOnly)(all.toVector filterNot { _ == library })
     val version = actualVersion(loader)(" (library jar  " + library.getAbsolutePath + ")")
     val compiler = compilerJar(scalaHome)
-    new ScalaInstance(version, loader, library, compiler, all.toArray, None)
+    new ScalaInstance(version, loader, loaderLibraryOnly, library, compiler, all.toArray, None)
```

Comment:
This is going to break quite a bit of code. I would honestly consider cutting a major breaking release... `xsbti.compile` is quite core to the Zinc API.

Reply:
> This is going to break quite a bit of code.

How so? Do you think there would be observable changes to sbt or other Zinc downstream users? (Besides being able to use the correct scala-xml version)

Reply:
Didn't mean behaviour-wise, but API-wise :smile:. I think `ScalaInstance` is used pretty much everywhere.

Reply:
I see what you mean. Maybe I can try to add another constructor that is bincompat and deprecate it.

Reply:
Yes, please do :+1:

Reply:
ok. Addressed the bincompat.